### ISSUE TITLE
[e2e] Umbrella issue for the e2e testing framework refactor

### ISSUE THREAD

**Issue by neolit123** [kubernetes/kubernetes#75601]
#### This is the umbrella issue for tracking work related to the refactoring of the e2e testing framework which is an effort by the [testing-commons group](https://docs.google.com/document/d/1TOC8vnmlkWw6HRNHoe5xSv5-qv7LelX6XK3UVCHuwb0/edit#) (#testing-commons) on slack - a sub-project of SIG testing.

[meeting VOD playlist](https://www.youtube.com/watch?v=7otFGfVB80w&list=PL69nYSiGNLP0ofY51bEooJ4TKuQtUSizR) (mixed with SIG testing meetings)

if you are working on PRs or filing tickets please reference this umbrella issues, **but don't close it** using `Fixes #...` or `Closes #...`. Also please don't reference the ticket in commit bodies.

please add the `[e2e]` prefix to your PR and tickets, but also label with `/area test-framework` and `/sig testing`.

Currently the framework suffers from a number of problems such as:
- complicated dependency graph (e.g k/k and cloud provider SDKs)
- internal vs external API client; the framework needs to use the external client-go client.
- ownership problems (who actively maintains tests in a SIG)
- architecture - providers need to be external and need to vendor the framework. provider tests should be external
- the framework uses non-incremental linking from the binary smashing it does. this process also requires libc.
- etc.

While the "framework" itself is located in `/test/e2e/framework`, in this tracking issues will can also cover work for `/test` in general.

this issue will include a number of high priority items that are tracked by the group. after a couple of meetings the list is the following:

------

**[P0] Decouple cloud providers from the framework.**
assigned: @andrewsykim @timothysc @neolit123
completed: TODO
tracking issue: https://github.com/kubernetes/kubernetes/issues/75604

details: tests that a part of the framework and the framework as a whole should not vendor cloud provider code. we need to think of a way how to interface this and possibly plan on how to bundle resulted test suite(s).

------

**[P1] Use the external client instead of internal**
assigned: @andrewsykim , @neolit123 
completed: yes
PRs:
- https://github.com/kubernetes/kubernetes/pull/75616
- https://github.com/kubernetes/kubernetes/pull/75989

details: the framework uses the internal client instead of client-go in all places, which creates a undesired dependency graph.

------

**[P2] Use the work from the WG component-standard to manage flags vs config.**
assigned: TODO
completed: TODO
tracking issue: TODO

details: currently the  number of flags that are compiled in the suite binary is overwhelming. there is also usage of the Viper library in there. there are a number of problems of how this is done. 
the WG in question is supposedly working on a solution for this problem - e.g. kflag.
we need to makes sure we don't break existing jobs that are defined in test-infra and pass flags to e2e.test.

------

**[P2/3] Refactor the framework utils.**

Separate tracking issue:
https://github.com/kubernetes/kubernetes/issues/76206

------

**[P3] Remove usage of IsMasterNode in the framework**
https://github.com/kubernetes/kubernetes/issues/80177

------

**[P3] Use e2elog of core e2e framework**

Separate tracking issue:
https://github.com/kubernetes/kubernetes/issues/81427

------

/sig testing
/sig cloud-provider
/sig cluster-lifecycle
(technically all/most SIGs have tests in the framework)
/area test
/kind design
/priority important longterm

/assign
/assign @timothysc  @andrewsykim @stevesloka @oomichi @pohly @alejandrox1

**Comment by neolit123** [comment]
/remove-kind bug
we are also looking at triaging the existing tickets and possibly cross reference.

@pohly i know you have a few tickets around.

@alejandrox1 @stevesloka 
i would look for tickets labeled `sig/testing` and `area/test` possibly keywords "framework".

**Comment by oomichi** [comment]
I'd like to join :-)

/assign @oomichi

**Comment by spiffxp** [comment]
WDYT about adding an `area/e2e-framework` label, and dropping an OWNERS file in test/e2e/framework that applies that label to PR's touching files in there?

**Comment by neolit123** [comment]
@spiffxp 
i think this will be useful.. today we've discussed the lack of specific label.

**Comment by spiffxp** [comment]
ok, I'm going to make a specific label then, and drop it in the repos that seem relevant to this effort, will link PR back here shortly

**Comment by neolit123** [comment]
https://github.com/kubernetes/kubernetes/pull/75611

**Comment by spiffxp** [comment]
Label may not show up until later today, waiting on label sync, work is in-flight on running that as a prowjob instead of a kubernetes-native CronJob

**Comment by neolit123** [comment]
thanks. this effort will require good amount of planning before the PRs and tickets start ramping up.

**Comment by neolit123** [comment]
list of the test/e2e deps after 1470df7a052eeed0 merged:
https://gist.github.com/neolit123/4798eba7addbe163324955751373bab3

current size of the suite is at 158MB.

**Comment by neolit123** [comment]
@alejandrox1 dug up a great refactor that we can apply:
https://github.com/kubernetes/kubernetes/issues/34059

the non-annotated assets are far from ideal.

**Comment by SataQiu** [comment]
I'm interested in this. 😆 
/assign @SataQiu

**Comment by soggiest** [comment]
Since there are so many open PRs associated with this umbrella can we move this to 1.16?

**Comment by neolit123** [comment]
@soggiest
yes this will span across multiple releases.
/milestone v1.16

**Comment by neolit123** [comment]
added new ticket to remove IsMasterNode:
https://github.com/kubernetes/kubernetes/issues/80177

help wanted!

**Comment by xmudrii** [comment]
@neolit123 Hello! I'm bug triage lead for the 1.16 release cycle and considering this issue is tagged for 1.16, but not updated for a long time, I'd like to check its status. The code freeze is starting on August 29th (about 1.5 weeks from now), which means that there should be a PR ready (and merged) until then.

Do we still target this issue to be fixed for 1.16?

**Comment by neolit123** [comment]
hi @xmudrii this ticket is tracking ongoing efforts that will span across multiple releases.
it contains works of `kind/cleanup` / `kind/design` mostly and not `kind/feature`, we should have it in the milestone as is, but update to 1.17 after 1.16 is out.

**Comment by alejandrox1** [comment]
Let's move it to the next milestone now. We are well into code freeze so most of this work will be blocked for the next week or so
/milestone v1.17

**Comment by neolit123** [comment]
hi @smourapina 
the work is still ongoing.
/milestone v1.18

**Comment by tanjunchen** [comment]
I'm interested in this. laughing
/assign @tanjunchen

**Comment by smourapina** [comment]
Hello @neolit123 and @tanjunchen ! 
Bug Triage team here for the 1.18 release. This is a friendly reminder that code freeze is scheduled for 5 March, which is about 3 weeks from now. Are you still targeting being ready before then, or would it be better if we move the milestone?

**Comment by alejandrox1** [comment]
Thank you Silvia for the reminder.
/milestone v1.19

**Comment by cheftako** [comment]
cc @jpbetz  @cici37  @nckturner @leilajal @jiahuif

**Comment by Danil-Grigorev** [comment]
cc @Danil-Grigorev

**Comment by spiffxp** [comment]
For those who have commented on this recently: what specifically are your plans?

The testing-commons subproject that was driving this was moved to best effort. If someone wants to continue working on this, that's great. Let the broader SIG Testing community know.

If nobody's driving a real plan though I am tempted to suggest we close this. Refactoring for refactoring's sake may not be the best use of our contributor and review bandwidth.

WDYT?

**Comment by spiffxp** [comment]
/remove-priority important-soon
/priority backlog
/lifecycle rotten
I'm starting the 30d clock to close this out unless someone steps forward to claim ownership
